### DISCUSSION TITLE
Auto Create Opening and Closing tag for HTML/JSX element in Zed?

### DISCUSSION THREAD

**Discussion by brijr** [zed-industries/zed#6666] [Feature Requests]
I love Zed and have switched from VSCode. One thing that has cost me a lot of time is not having the feature where when you type a HTML or JSX tag ie. `h2` the editor automatically creates the opening and closing tag for that HTML element ie. `<h2></h2>`.

Am I just missing the setting? Or is this not built into Zed?

**Comment by JosephTLyons** [comment]
You are right, we currently don't have this, but it feels so basic and essential - we will add this in, either as a raw autocompletion, or via Emmet, after we go open source.

**Reply by brijr** [reply]
Amazing! Thanks for the response. All the best!

**Reply by rossirpaulo** [reply]
Any updates on this?

**Reply by brijr** [reply]
@rossirpaulo I don't believe they have gone Open Source yet.

**Reply by rossirpaulo** [reply]
Oh, so they are only fixing it on open source? It seems like a straightforward fix!

**Reply by rossirpaulo** [reply]
Love the app too, but it really does increase overhead drastically :(

**Reply by brijr** [reply]
@rossirpaulo Agreed. Yeah that's what I got out of it that they wouldn't update til OS for some reason?

**Reply by JosephTLyons** [reply]
We are very close to open souce status. Our theory is that opening the code up is the number one thing we can to improve the quality of Zed, because the entire Zed community can help out if they'd like to.  Once the community has access, I think tons of issues will be getting fixed - it is a hard choice, but it feels like the right move, to not keep pushing our internal date out further. We are still fixing critical global issues, but per-language ones have all sort of fallen to the side, since there are a large pool of them. Hang tight.

I'm tracking where we are in our plans here, we have progress on this issue! :)

- https://github.com/zed-industries/community/issues/2197

**Reply by joaopcm** [reply]
Is the auto closing tags for JSX/TSX already implemented? It doesn't seem to be working for me :/

**Reply by Haume0** [reply]
not working for me either

**Reply by kdrai007** [reply]
not working for me either.
is there any extension or something that can fix this issue

**Reply by dimasyrnyk** [reply]
Try to install extension "Emmet". It works for me if I type tag name and press "tab" button.

**Reply by sidvud98** [reply]
> Try to install extension "Emmet". It works for me if I type tag name and press "tab" button.

Thanks! Works...

**Reply by mico-fwen** [reply]
Doesn't work for JSX/TSX at the moment.

**Reply by LUISEDOCCOR** [reply]
> Try to install extension "Emmet". It works for me if I type tag name and press "tab" button.

Thanks! I had the same problem but installing the Emmet extension solved it.

**Reply by itsanji** [reply]
> Try to install extension "Emmet". It works for me if I type tag name and press "tab" button.

thanks, work for me!!

**Reply by stpn48** [reply]
> Try to install extension "Emmet". It works for me if I type tag name and press "tab" button.

thanks bro works for me. I really needed that, its so annoying to type the closing tag.

**Reply by 06kellyjac** [reply]
@mico-fwen if you're using the emmet extension make sure you only type `div` then press tab, not `<div` or `<div>` and it should just work :)

And don't forget the original power of emmet doing stuff like `div.myclass` -> `<div className="myclass"></div>` (in jsx/tsx)

`div{hello}.myclass-$*5#myid` ->

```
<div className="myclass-1" id="myid">hello</div>
<div className="myclass-2" id="myid">hello</div>
<div className="myclass-3" id="myid">hello</div>
<div className="myclass-4" id="myid">hello</div>
<div className="myclass-5" id="myid">hello</div>
```

**Comment by jawade-arpeet** [comment]
https://github.com/user-attachments/assets/ae9fddc2-42f0-4e43-981d-dd9a5155f5f9

It would be much better if it worked like this in webstorm instead of typing tag name and pressing "tab" button.

**Reply by Haume0** [reply]
thats good too but writing "div" and pressing enter/tab its much faster way to write code.

**Reply by pradsgit** [reply]
Is this still not available? I am unable to make it work

**Reply by jawade-arpeet** [reply]
No, but you can install emmet and use tab as @dimasyrnyk  mentioned.

**Reply by ChristoferMendes** [reply]
> thats good too but writing "div" and pressing enter/tab its much faster way to write code.

Yes, but not so fast with imported components in React. Because opening the componentv with (<) Zed suggests imports, while not opening the tag, Zed will not suggest it

**Reply by dmytrosyrnyk** [reply]
The problem is not only in speed. It doesn't work for tags that contains one letter like b,p,a, etc.

**Reply by knownIndie** [reply]
Any update on this issue

**Comment by aretrace** [comment]
Any timeline for this?

**Comment by albingroen** [comment]
Yeah I kind of need this. For example, I work a lot in React and React Native, and sure Emmet works when using built-in tags like div, p, h1, etc. But when I want to write `<Typography.Paragraph>`, I expect zed to give me the possibility to auto import the component as well as auto close the tag with `</Typography.Paragraph>`. If someone adds this I'm probably moving over from Neovim :)

https://github.com/user-attachments/assets/1ec71e0c-ad2b-4009-9627-591da85c5b66

**Comment by umkl** [comment]
Any updates on this?

**Comment by mfsalis** [comment]
Any updates? 2 years has passed..

**Reply by dimasyrnyk** [reply]
Just use the extension "Emmet", now it's working even with one letter elements like a,b,i,p, etc. as well as with components.
